xdsl/dialects/fsm.py

```diff
@@ -0,0 +1,379 @@
+from __future__ import annotations
+
+from collections.abc import Sequence
+
+from xdsl.dialects.arith import signlessIntegerLike
+from xdsl.dialects.builtin import (
+    ArrayAttr,
+    DictionaryAttr,
+    FlatSymbolRefAttr,
+    FunctionType,
+    StringAttr,
+)
+from xdsl.ir import (
+    Attribute,
+    Block,
+    Operation,
+    Region,
+    SSAValue,
+)
+from xdsl.ir.core import Dialect
+from xdsl.irdl import (
+    AnyAttr,
+    IRDLOperation,
+    attr_def,
+    irdl_op_definition,
+    operand_def,
+    opt_attr_def,
+    opt_operand_def,
+    region_def,
+    var_operand_def,
+    var_result_def,
+)
+from xdsl.traits import (
+    HasParent,
+    IsTerminator,
+    NoTerminator,
+    SymbolOpInterface,
+    SymbolTable,
+)
+from xdsl.utils.exceptions import VerifyException
+
+"""
+This file defines the FSM dialect, a CIRCT dialect for FSM description and manipulation.
+https://circt.llvm.org/docs/Dialects/FSM/
+"""
+
+
+@irdl_op_definition
+class TransitionOp(IRDLOperation):
+    """
+    Represents a transition of a state with a symbol reference
+    of the next state. This op includes an optional `$guard` region with an `fsm.return`
+    as terminator that returns a Boolean value indicating the guard condition of
+    this transition. This op also includes an optional `$action` region that represents
+    the actions to be executed when this transition is taken
+    """
+
+    name = "fsm.transition"
+
+    guard = region_def()
+
+    action = region_def()
+
+    nextState = attr_def(FlatSymbolRefAttr)
+
+    def __init__(
+        self,
+        nextState: FlatSymbolRefAttr,
+        guard: Region | type[Region.DEFAULT] = Region.DEFAULT,
+        action: Region | type[Region.DEFAULT] = Region.DEFAULT,
+    ):
+        if isinstance(nextState, str):
+            nextState = FlatSymbolRefAttr(nextState)
+        attributes: dict[str, Attribute] = {
+            "nextState": nextState,
+        }
+        if not isinstance(action, Region):
+            action = Region(Block())
+        if not isinstance(guard, Region):
+            guard = Region(Block())
+        super().__init__(
+            attributes=attributes,
+            regions=[guard, action],
+        )
+
+    def verify_(self):
+        if SymbolTable.lookup_symbol(self, self.nextState) is None or not isinstance(
```

Comment:
I would probably create a function for `SymbolTable.lookup_symbol(self, self.nextState)` (since it's going to be useful later on), and I would make this a variable, so you don't lookup the symbol twice.